crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -7624,10 +7624,15 @@ impl<'db> TypeInferenceBuilder<'db> {
             match element {
                 ast::Expr::EllipsisLiteral(_) | ast::Expr::Starred(_) => true,
                 ast::Expr::Subscript(ast::ExprSubscript { value, .. }) => {
-                    matches!(
-                        builder.expression_type(value),
-                        Type::KnownInstance(KnownInstanceType::Unpack)
-                    )
+                    let value_ty = if builder.deferred_state.in_string_annotation() {
+                        // Using `.expression_type` does not work in string annotations, because
+                        // we do not store types for sub-expressions. Re-infer the type here.
+                        builder.infer_expression(value)
+                    } else {
+                        builder.expression_type(value)
+                    };
```

Comment:
oof, that seems a bit unfortunate. Does that mean we should just get rid of the `expression_type()` API altogether? Or have `expression_type()` itself check whether we're in a deferred string annotation (and fallback to `infer_expression()` if so)?

It feels like a bit of a footgun if we have to always make sure we're not in a deferred string annotation before calling `expression_type()`

Reply:
Good questions. It's certainly possible that we have other hidden problems with calling `expression_type` for something that might be inside a string annotation. But note that it's not common to re-enter a sub-AST and match on inner elements (and then use `expression_type`) after the type has already been inferred. Usually, in functions like `infer_expression`, we return types "back up" to the caller, i.e. to outer elements. Admittedly, I did not think too much about this exact callsite because this whole function is basically a large "To do" and will disappear eventually.